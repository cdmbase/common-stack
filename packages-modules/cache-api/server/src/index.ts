import module from './module';
import { Feature } from '@common-stack/server-core';

export * from './constants';
export { Cache } from './services';
export default new Feature(module);

