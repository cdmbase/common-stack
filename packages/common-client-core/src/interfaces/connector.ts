// tslint:disable:max-line-length

import { LocalStateFragmentMatcher } from 'apollo-client';
import { IdGetter } from 'apollo-cache-inmemory';
import { ReducersMapObject } from 'redux';
import { ErrorLink } from 'apollo-link-error';
export interface IClientStateConfig {
    resolvers?: any; // don't need `Resolvers` type as it may conflict with the usage
    defaults?: any;
    typeDefs?: string | string[];
    fragmentMatcher?: LocalStateFragmentMatcher;
}


/**
 * ModuleShape have optional configuration to be implemented by all the feature modules
 * in the application.
 */
export interface IModuleShape {
    /**
     * @param link all `apollo-link` that need to be composed.
     * @inheritdoc https://github.com/apollographql/apollo-link
     */
    readonly link?: any;
    /**
     * @param errorLink compose all errorLink
     * @inheritdoc https://github.com/apollographql/apollo-link/tree/master/packages/apollo-link-error
     */
    readonly errorLink?: ErrorLink | ErrorLink[];
    readonly createFetch?: any;
    /**
     * @param connectionParam method that called when a client connects to the socket.
     * @inheritdoc https://github.com/apollographql/subscriptions-transport-ws#constructorurl-options-connectioncallback
     */
    readonly connectionParam?: any;
    /**
     * @param epic all `epics` that need to be composed.
     * @inheritdoc https://redux-observable.js.org/docs/basics/Epics.html
     */
    readonly epic?: any;
    /**
     * @deprecated not sure we using
     */
    readonly reduxContext?: any;
    /**
     * @param reducer Redux reducers list
     */
    readonly reducer?: ReducersMapObject | ReducersMapObject[];
    /**
     * @param clientStateParams Client side configuration of `apollo-client`
     * @inheritdoc https://github.com/apollographql/apollo-client/tree/2d65da133c156f6d808e64aee7e7fd5f7cc71d7f
     */
    readonly clientStateParams?: IClientStateConfig | IClientStateConfig[];
    /**
     * @param dataIdFromObject A function that returns an object identifier given an particular result
     * object.
     * @inheritdoc https://github.com/apollographql/apollo-client/tree/ed66999bac40226abfeada8d6c83b454636bb4b0/packages/apollo-cache-inmemory#configuration
     */
    readonly dataIdFromObject?: { [key: string]: IdGetter } |  { [key: string]: IdGetter }[];
    /**
     * @param createContainerFunc Synchronous Container Modules of inversify.
     * @inheritdoc https://github.com/inversify/InversifyJS/blob/master/wiki/container_modules.md
     */
    readonly createContainerFunc?: Function | Function[];
    readonly sidebarSegments?: any;
    readonly routerFactory?: any;
    /**
     * @param route Route list
     */
    readonly route?: any;
    /**
     * @param routeConfig Custom route data to create React Routers
     */
    readonly routeConfig?: any;
    /**
     * @param menuConfig Menu data to create side menu
     */
    readonly menuConfig?: any;
    /**
     * @param navItem Top left navigation links
     */
    readonly navItem?: any;
    /**
     * @param navItemRightTop right navigation links
     */
    readonly navItemRight?: any;
    /**
     * @param rootComponentFactory Root component factory list
     */
    readonly rootComponentFactory?: any;
    /**
     * @param dataRootComponent Data root React elements list (data root elements wraps data fetching react subtree root)
     */
    readonly dataRootComponent?: any;
    readonly createFetchOptions?: any;
    /**
     * @param stylesInsert URL list to 3rd party css scripts
     */
    readonly stylesInsert?: any;
    /**
     * @param scriptsInsert URL list to 3rd party js scripts
     */
    readonly scriptsInsert?: any;
    /**
     * @deprecated
     */
    readonly catalogInfo?: any;
    /**
     * @deprecated
     */
    readonly languagesFuncs?: any;
    /**
     * @param leftMainPanelItems Components that will be placed left panel.
     */
    readonly leftMainPanelItems?: any;
    /**
     * @param middleMainPanelItems Components that will be placed middle center panel
     */
    readonly middleMainPanelItems?: any;
    /**
     * @param middleMainPanelItemsProps Props of the component that will be placed middle center panel
     */
    readonly middleMainPanelItemsProps?: any;
    /**
     * @param leftFooterItems Components that will be placed left footer panel
     */
    readonly leftFooterItems?: any;
    /**
     * @param rightFooterItems Components that will be placed right footer panel
     */
    readonly rightFooterItems?: any;
    /**
     * @param middleLowerPanelItems Components that will be placed below middle panel
     */
    readonly middleLowerPanelItems?: any;

}

export interface IFeature extends IModuleShape {
    // Public variables
    readonly data: any[];

    /**
     * @returns Redux-Observable Epics
     */
    readonly epics;

    readonly getRouter;
    /**
     * @returns client-side React route components list
     */
    readonly getRoutes;
    readonly getConfiguredRoutes;

    readonly getMenus;
    readonly getConfiguredMenus;
    /**
     * @returns client-side top left navbar link component list
     */
    readonly navItems;
    /**
     * @returns client-side top right navbar link component list
     */
    readonly navItemsRight;
    /**
     * @returns Redux reducers
     */
    readonly reducers;

    readonly connectionParams;

    /**
     * @returns URL list to 3rd party css styles
     */
    readonly stylesInserts;
    /**
     * @returns URL list to 3rd party js styles
     */
    readonly scriptsInserts;

    readonly leftMainPanel;
    readonly middleMainPanel;
    readonly leftFooter;
    readonly rightFooter;
    readonly middleLowerPanel;
    readonly dataIdFromObject;

    /**
     * @param args Options to pass to each Container Module
     * @returns Created container and binds all the container modules
     */
    createContainers(args);
    /**
     * @param args Provide resolverContext which can be passed to all resolver functions.
     * @returns IClientStateConfig
     */
    getStateParams(args?: {resolverContex?: any }): IClientStateConfig;
    /**
     * @arguments root React tree root component
     * @arguments req Http Request
     * @returns React tree root component wrapped up by root components exposed by this module
     */
    getWrappedRoot(root: any, req?: any): any;
    /**
     * @param root React tree data root component (first React root components which is used for fetching data)
     *
     * @returns React tree data root component wrapped up by data root components exposed by this module
     */
    getDataRoot(root: any): any;
    registerLanguages(monaco): void;
}
