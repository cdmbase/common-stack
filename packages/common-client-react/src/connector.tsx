import * as React from 'react';

import { AbstractFeature, IFeature } from '@common-stack/client-core';


export class Feature extends AbstractFeature implements IFeature {
    get routes() {
        return this.route.map((component, idx) => React.cloneElement(component, { key: idx + this.route.length }));
    }

    get navItems() {
        return this.navItem.map((component, idx) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    get navItemsRight() {
        return this.navItemRight.map((component, idx) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    public getWrappedRoot(root, req) {
        let nestedRoot = root;
        for (const componentFactory of this.rootComponentFactory) {
            nestedRoot = React.cloneElement(componentFactory(req), {}, nestedRoot);
        }
        return nestedRoot;
    }

}
