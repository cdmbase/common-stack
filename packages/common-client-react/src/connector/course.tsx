import * as React from 'react';
import { Link } from 'react-router-dom';

interface MyProps {
  history?: any;
  children?: any;
  key?: any;
};

const Course: React.StatelessComponent<MyProps> = (props: MyProps) => {
  const from = "localhost:3000/"; 
  return (
  <div key={props.history.id}>
    <Link to={`${props.history.number}`} {...from}>{`Page: ${
        props.history.number
      }`}</Link>
  </div>
)};

export default Course;