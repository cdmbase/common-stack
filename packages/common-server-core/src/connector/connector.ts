import { DocumentNode } from 'graphql';

import { merge, map, union, without, castArray } from 'lodash';

export const featureCatalog: any = {};

const combine = (features, extractor): any =>
  without(union(...map(features, res => castArray(extractor(res)))), undefined);

export type FeatureParams = {
  schema?: DocumentNode | DocumentNode[],
  createDirectivesFunc?: Function | Function[],
  createResolversFunc?: Function | Function[],
  createContextFunc?: Function | Function[],
  beforeware?: any | any[],
  middleware?: any | any[],
  createFetchOptions?: Function | Function[],
  catalogInfo?: any | any[],
};

class Feature {
  public schema: DocumentNode[];
  public createDirectivesFunc: Function[];
  public createResolversFunc: Function[];
  public createContextFunc: Function[];
  public createFetchOptions: Function[];
  public beforeware: Function[];
  public middleware: Function[];

  constructor(feature?: FeatureParams, ...features: Feature[]) {
    // console.log(feature.schema[0] instanceof DocumentNode);
    combine(arguments, arg => arg.catalogInfo).forEach(info =>
      Object.keys(info).forEach(key => (featureCatalog[key] = info[key])),
    );
    this.schema = combine(arguments, arg => arg.schema);
    this.createDirectivesFunc = combine(arguments, arg => arg.createDirectivesFunc);
    this.createResolversFunc = combine(arguments, arg => arg.createResolversFunc);
    this.createContextFunc = combine(arguments, arg => arg.createContextFunc);
    this.beforeware = combine(arguments, arg => arg.beforeware);
    this.middleware = combine(arguments, arg => arg.middleware);
    this.createFetchOptions = combine(arguments, arg => arg.createFetchOptions);
  }

  get schemas(): DocumentNode[] {
    return this.schema;
  }

  public async createContext(req: any, connectionParams: any, webSocket?: any) {
    const results = await Promise.all(
      this.createContextFunc.map(createContext => createContext(req, connectionParams, webSocket)),
    );
    return merge({}, ...results);
  }

  public createResolvers(pubsub: any) {
    return merge({}, ...this.createResolversFunc.map(createResolvers => createResolvers(pubsub)));
  }

  public createDirectives() {
    return merge({}, ...this.createDirectivesFunc.map(createDirectives => createDirectives()));
  }

  get beforewares(): any[] {
    return this.beforeware;
  }

  get middlewares(): any[] {
    return this.middleware;
  }

  get constructFetchOptions(): any {
    return this.createFetchOptions.length
      ? (...args) => {
        try {
          let result = {};
          for (let func of this.createFetchOptions) {
            result = { ...result, ...func(...args) };
          }
          return result;
        } catch (e) {
          console.log(e.stack);
        }
      }
      : null;
  }
}

export { Feature };
