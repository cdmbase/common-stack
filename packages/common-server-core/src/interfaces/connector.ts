import * as Logger from 'bunyan';

export interface IResolverOptions {
    pubsub: any;
    logger?: Logger;
    [key: string]: any;
}

export interface IDirectiveOptions {
    [key: string]: any;
    logger?: Logger;
}

export enum ConfigurationScope {
    WINDOW = 1,
    RESOURCE,
}

export interface IWebsocketConfig {
    [key: string]: any;
}

export interface IPreferences<T = ConfigurationScope> {
    [key: string]: IPreferencesData<T>;
}

export interface IRoles<T = ConfigurationScope> {
    [key: string]: IPreferencesData<T>;
}
export interface IPreferencesData<T = ConfigurationScope> {
    type?: string | string[];
    default?: string | boolean | number | any;
    description?: string;
    overridable?: boolean;
    scope?: T;
    settings?: string;
    enum?: any;
    enumDescriptions?: string[];
    [key: string]: any;
}

/**
 * Formatted preferences data
 */
export interface IPreferncesTransformed<T = ConfigurationScope> {
    type: string;
    data: IPreferencesData<T>;
}

export interface IOverwritePreference {
    [key: string]: string | boolean | number | any;
}
