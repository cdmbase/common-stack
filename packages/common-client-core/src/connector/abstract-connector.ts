import { IFeature } from '../interfaces';
import { merge, map, union, without, castArray } from 'lodash';

const combine = (features, extractor) => without(union(...map(features, res => castArray(extractor(res)))), undefined);
export const featureCatalog = {};

export abstract class AbstractFeature implements IFeature {
    protected route;
    protected navItem;
    protected navItemRight;
    protected reducer;
    protected resolver;
    protected middleware;
    protected afterware;
    protected connectionParam;
    protected createFetchOptions;
    protected stylesInsert;
    protected scriptsInsert;
    protected rootComponentFactory;
    protected routerFactory;

    constructor(
        {
          route,
            navItem,
            navItemRight,
            reducer,
            resolver,
            middleware,
            afterware,
            connectionParam,
            createFetchOptions,
            stylesInsert,
            scriptsInsert,
            rootComponentFactory,
            routerFactory,
            catalogInfo,
        },
        ...features,
    ) {
        /* eslint-enable no-unused-vars */
        combine(arguments, arg => arg.catalogInfo).forEach(info =>
            Object.keys(info).forEach(key => (featureCatalog[key] = info[key]))
        );
        this.route = combine(arguments, arg => arg.route);
        this.navItem = combine(arguments, arg => arg.navItem);
        this.navItemRight = combine(arguments, arg => arg.navItemRight);
        this.reducer = combine(arguments, arg => arg.reducer);
        this.resolver = combine(arguments, arg => arg.resolver);
        this.middleware = combine(arguments, arg => arg.middleware);
        this.afterware = combine(arguments, arg => arg.afterware);
        this.connectionParam = combine(arguments, arg => arg.connectionParam);
        this.createFetchOptions = combine(arguments, arg => arg.createFetchOptions);
        this.stylesInsert = combine(arguments, arg => arg.stylesInsert);
        this.scriptsInsert = combine(arguments, arg => arg.scriptsInsert);
        this.rootComponentFactory = combine(arguments, arg => arg.rootComponentFactory);
        this.routerFactory = combine(arguments, arg => arg.routerFactory)
            .slice(-1)
            .pop();
    }

    public router() {
        return this.routerFactory();
    }

    public abstract routes();

    public abstract navItems();

    public abstract navItemsRight();

    public reducers() {
        return merge(...this.reducer);
    }

    public resolvers() {
        return merge(...this.resolver);
    }

    public middlewares() {
        return this.middleware;
    }

    public afterwares() {
        return this.afterware;
    }

    public connectionParams() {
        return this.connectionParam;
    }

    public constructFetchOptions() {
        return this.createFetchOptions.length
            ? (...args) => {
                try {
                    let result = {};
                    for (let func of this.createFetchOptions) {
                        result = { ...result, ...func(...args) };
                    }
                    return result;
                } catch (e) {
                    console.log(e.stack);
                }
            }
            : null;
    }

    public stylesInserts() {
        return this.stylesInsert;
    }

    public scriptsInserts() {
        return this.scriptsInsert;
    }

    public abstract getWrappedRoot(root, req);
}
