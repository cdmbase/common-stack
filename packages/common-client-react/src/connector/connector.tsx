import * as React from 'react';
import {Route, Switch} from 'react-router-dom';
import {AbstractFeature} from '@common-stack/client-core';
import * as Logger from 'bunyan';
import {IPlugin, IReactFeature} from '../interfaces';
import {getMenus, getRoutes} from '../utils';
import {castArray, map, union, without} from 'lodash';
import {registerPlugin, getPlugins, getPlugin} from '../plugin-area/plugin-api';

const combine = (features, extractor) => without(union(...map(features,
        res => castArray(extractor(res)))), undefined);

export class Feature extends AbstractFeature implements IReactFeature {
    private logger;
    public componentFillPlugins;
    constructor(
        feature?: IReactFeature,
        // tslint:disable:trailing-comma
        ...features: IReactFeature[]
    ) {
        super(feature, ...features);
        this.logger = Logger.createLogger({
            name: Feature.name
        });
        this.componentFillPlugins = this.registerComponentFillPlugins(combine(arguments,
            (arg: IReactFeature) => arg.componentFillPlugins));
    }

    /* tslint:disable:jsx-no-lambda */
    private renderRoutes = (routes, solidRoutes, extraProps = {}, switchProps = {}) =>
        routes ? (
            <Switch {...switchProps}>
                {[
                    ...solidRoutes, ...routes.map((route, i) => (
                        <Route
                            key={route.key || i}
                            path={route.path}
                            exact={route.exact}
                            strict={route.strict}
                            render={props => (
                                <route.component {...props} {...extraProps} route={route}/>
                            )}
                        />
                    )),
                ]}
            </Switch>
        ) : null

    /**
     * Get the routes
     */
    public getRoutes(withRoot?: boolean, rootComponent?: any) {
        const configuredRoutes = this.getConfiguredRoutes();
        const solidRoutes = this.route.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {key: component.props.path}));
        return this.renderRoutes(configuredRoutes, solidRoutes);
    }

    /**
     * Get menus
     */
    public getMenus(withRoot?: boolean, rootComponent?: any) {
        return this.getConfiguredMenus();
    }

    /**
     * get configured routes.
     * Note: It overwrites the any duplicate key with latest loaded key.
     * TODO: Find a way to warn when there are duplicate keys.
     */
    public getConfiguredRoutes(searchPath = '/') {
        const routes = Object.assign({}, ...this.routeConfig);
        return getRoutes(searchPath, {...routes});
    }


    /**
     * get configured menus.
     * Note: It overwrites the any duplicate key with latest loaded key.
     * TODO: Find a way to warn when there are duplicate keys.
     */
    public getConfiguredMenus(searchPath = '/') {
        const routes = Object.assign({}, ...this.menuConfig);
        return getMenus(searchPath, {...routes});
    }

    get navItems() {
        return this.navItem.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItem.length,
            }),
        );
    }

    get navItemsRight() {
        return this.navItemsRight.map((component: React.ReactElement<any>, idx: number) =>
            React.cloneElement(component, {
                key: component.key ? component.key : idx + this.navItems.length,
            }),
        );
    }

    public getWrappedRoot(root: React.ReactNode, req?: any): React.ReactNode {
        let nestedRoot = root;
        for (const componentFactory of this.rootComponentFactory) {
            nestedRoot = React.cloneElement(componentFactory(req), {}, nestedRoot);
        }
        return nestedRoot;
    }

    public getDataRoot(root) {
        let nestedRoot = root;
        for (const component of this.dataRootComponent) {
            nestedRoot = React.createElement(component, {}, nestedRoot);
        }
        return nestedRoot;
    }

    public registerLanguages(monaco) {
        for (const func of this.languagesFuncs) {
            func(monaco);
        }
    }

    private registerComponentFillPlugins(plugins): void {
        plugins.forEach((i) => {
            const {name} = i;
            const isFound = !!getPlugin(name);
            if (!isFound) {
                registerPlugin(name, i);
            }
        });
        return plugins;
    }

    public getComponentFillPlugins() {
        return getPlugins();
    }
}
