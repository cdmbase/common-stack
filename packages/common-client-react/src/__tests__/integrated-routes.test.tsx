import * as React from 'react';
import * as ReactDOMServer from 'react-dom/server';
import StaticRouter from 'react-router/StaticRouter';
import { renderRoutes, matchRoutes } from 'react-router-config';
import { Feature } from '../connector';
import { ConsoleLogger, IConsoleLoggerSettings } from '@cdm-logger/server';
import { CdmLogger } from '@cdm-logger/core';
const logger: CdmLogger.ILogger = ConsoleLogger.create('test', { level: 'trace' });
import 'jest';

describe('integration', () => {
    it('generates the same matches in renderRoutes and matchRoutes', () => {
        const rendered = [];

        const Comp = ({ match, route: { routes } }) => (
            rendered.push(match), renderRoutes(routes)
        );

        const genRoutes = [
            {
                ['/ghost']: { component: Comp },
                ['/pepper']: { component: Comp, exact: false },
                ['/pepper/:type']: { component: Comp,  exact: false},
                ['/pepper/:type/scoville']: { component: Comp, exact: false },
            },
        ];
        const result = [
            {
                path: '/ghost',
                component: Comp,
            },
            {
                path: '/pepper',
                component: Comp,
                exact: false,
                routes: [
                    {
                        path: '/pepper/:type',
                        component: Comp,
                        exact: false,
                        routes: [
                            {
                                path: '/pepper/:type/scoville',
                                component: Comp,
                                exact: false,
                            },
                        ],
                    },
                ],
            },
        ];
        const connector = new Feature({ routeConfig: genRoutes });
        const connectorRoutes = connector.getConfiguredRoutes();
        expect(connectorRoutes).toMatchObject(result);
        const pathname = '/pepper/jalepeno';
        const branch = matchRoutes(connectorRoutes, pathname);
        ReactDOMServer.renderToString(
            <StaticRouter location={pathname} context={{}}>
                {renderRoutes(connectorRoutes)}
            </StaticRouter>,
        );

        expect(branch).toMatchSnapshot();
        expect(branch.length).toEqual(2);
        expect(rendered.length).toEqual(2);
        expect(branch[0].match).toEqual(rendered[0]);
        expect(branch[1].match).toEqual(rendered[1]);
    });

});
