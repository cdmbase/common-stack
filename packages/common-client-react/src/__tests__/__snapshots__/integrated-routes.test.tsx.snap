// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration generates the same matches in renderRoutes and matchRoutes 1`] = `
Array [
  Object {
    "match": Object {
      "isExact": false,
      "params": Object {},
      "path": "/pepper",
      "url": "/pepper",
    },
    "route": Object {
      "component": [Function],
      "exact": false,
      "path": "/pepper",
      "routes": Array [
        Object {
          "component": [Function],
          "exact": false,
          "path": "/pepper/:type",
          "routes": Array [
            Object {
              "component": [Function],
              "exact": false,
              "path": "/pepper/:type/scoville",
            },
          ],
        },
      ],
    },
  },
  Object {
    "match": Object {
      "isExact": true,
      "params": Object {
        "type": "jalepeno",
      },
      "path": "/pepper/:type",
      "url": "/pepper/jalepeno",
    },
    "route": Object {
      "component": [Function],
      "exact": false,
      "path": "/pepper/:type",
      "routes": Array [
        Object {
          "component": [Function],
          "exact": false,
          "path": "/pepper/:type/scoville",
        },
      ],
    },
  },
]
`;
